src/controllers/scripting/controllerscriptenginebase.cpp

```diff
@@ -0,0 +1,250 @@
+#include "controllers/scripting/controllerscriptenginebase.h"
+
+#include "control/controlobject.h"
+#include "controllers/controller.h"
+#include "controllers/controllerdebug.h"
+#include "controllers/scripting/colormapperjsproxy.h"
+#include "controllers/scripting/legacy/controllerscriptinterface.h"
+#include "errordialoghandler.h"
+#include "mixer/playermanager.h"
+
+ControllerScriptEngineBase::ControllerScriptEngineBase(Controller* controller)
+        : m_bDisplayingExceptionDialog(false),
+          m_pJSEngine(nullptr),
+          m_pController(controller),
+          m_bTesting(false) {
+    // Handle error dialog buttons
+    qRegisterMetaType<QMessageBox::StandardButton>("QMessageBox::StandardButton");
+
+    // Subclasses are responsible for adding paths to watch to m_scriptWatcher
+    // in ther initializeScriptEngine method.
+    connect(&m_scriptWatcher,
+            &QFileSystemWatcher::fileChanged,
+            this,
+            &ControllerScriptEngineBase::reload);
+}
+
+ControllerScriptEngineBase::~ControllerScriptEngineBase() {
+    shutdown();
+}
+
+bool ControllerScriptEngineBase::initialize() {
+    VERIFY_OR_DEBUG_ASSERT(!m_pJSEngine) {
+        return false;
+    }
+
+    // Create the Script Engine
+    m_pJSEngine = new QJSEngine(this);
+
+    QJSValue engineGlobalObject = m_pJSEngine->globalObject();
+
+    QJSValue mapper = m_pJSEngine->newQMetaObject(
+            &ColorMapperJSProxy::staticMetaObject);
+    engineGlobalObject.setProperty("ColorMapper", mapper);
+
+    if (m_pController) {
+        qDebug() << "Controller in script engine is:"
+                 << m_pController->getName();
+
+        ControllerJSProxy* controllerProxy = m_pController->jsProxy();
+
+        // Make the Controller instance available to scripts
+        engineGlobalObject.setProperty(
+                "controller", m_pJSEngine->newQObject(controllerProxy));
+
+        // ...under the legacy name as well
+        engineGlobalObject.setProperty(
+                "midi", m_pJSEngine->newQObject(controllerProxy));
+    }
+
+    m_byteArrayToScriptValueJSFunction = m_pJSEngine->evaluate(
+            "(function(arg1) { return new Uint8Array(arg1) })");
+
+    return true;
+}
+
+void ControllerScriptEngineBase::shutdown() {
+    m_scriptWatcher.removePaths(m_scriptWatcher.directories());
```

Comment:
I don't understand how this is inconsistent. If this wasn't here I think this line would need to be duplicated between `reload` and the destructor.

Reply:
The watcher is installed in the constructor and uninstalled in shut down. After reload it is gone.

Reply:
This is to avoid reallocating a new QFileSystemWatcher when reloading. Only the paths need to be cleared from the watcher, but this avoid deleting and creating a new object.

Reply:
I'm in doubt that works. If so, please make it explicit somehow.

Reply:
If you want I can add a comment to clarify why this is only removing the path and not deleting.